
# -------------------------------------------------------------------------------------
#
#
#    Script for running LDSC on snRNA-seq data - on SCZ GWAS only at this point
#
#
# -------------------------------------------------------------------------------------

# ---------  SET SMK PARAMS  ----------
configfile: "config/config.yaml"
report: "report/workflow.rst" # Experimented with this but not currently using

# ----------  SET VARIABLES  ----------
SCRATCH = "/scratch/c.c1477909/"
LDSC_DIR = SCRATCH + "ldsc/"
REF_DIR = LDSC_DIR + "reference_files/"
RESULTS_DIR = SCRATCH + "results/"
MARKDOWN_DIR = SCRATCH + "markdown/"
MAGMA_DIR = SCRATCH + "magma_celltyping/"
ANN_DIR = RESULTS_DIR + "LDSR_annotation_files/RNA/"
PART_HERIT_DIR = RESULTS_DIR + "LDSR_part_herit_files/RNA/"
GWAS_DIR = LDSC_DIR + "GWAS_for_ldsc/"
QUANTILES_DIR = MAGMA_DIR + "ldsc_gene_lists/"


# -------------  RULES  ---------------

# - https://github.com/bulik/ldsc/wiki/LD-Score-Estimation-Tutorial

# - make_annot - make ldsc annotation files - 220 files per cell type - one per quantile per chromosome - 1 x 11 x 22 = 220 files 
#      - gene set file - generated by snRNAseq_magma_create_LDSC_geneLists.R
#      - gene coord files - coordinates for all protein coding genes - accessed here http://grch37.ensembl.org/biomart - modified using snRNAseq_LDSC_get_gene_coords.R 
# - ldsr - caculate ld scores - 220 files per cell type 
# - partitioned heritability - calculate part herit enrichments- 11 files per cell type per GWAS 

rule all:
    input:
         expand(PART_HERIT_DIR + "snRNAseq_LDSC_{CELL_TYPE}_{GWAS}_summary.tsv", CELL_TYPE = config["RNA_CELL_TYPES"], GWAS = config["LDSC_GWAS"])

rule make_annot:
    input:   gene_set = QUANTILES_DIR + "{CELL_TYPE}_Q{QUANTILE}_genes.tsv",
             gene_coord = LDSC_DIR + "snRNAseq_LDSC_gene_coords.tsv",
             bim_file = REF_DIR + "1000G_EUR_Phase3_plink/1000G.EUR.QC.{CHR}.bim" 
    output:  ANN_DIR + "snRNAseq.{CELL_TYPE}.Q{QUANTILE}.{CHR}.annot.gz"
    conda:   "envs/ldsc.yml"    
    message: "Creating annotation files for snRNAseq: {wildcards.CELL_TYPE} Quantile {wildcards.QUANTILE}, Chr {wildcards.CHR} "
    log:     "logs/ldsc/make_annot.snRNAseq.{CELL_TYPE}.Q{QUANTILE}.Chr{CHR}.log"
    shell:
        """
        
        python ldsc/make_annot.py \
        --gene-set-file {input.gene_set} \
        --gene-coord-file {input.gene_coord} \
        --windowsize 100000 \
        --bimfile {input.bim_file} \
        --annot-file {output} 2> {log} \
        
        """

rule ldsr:
    input:   annot = ANN_DIR + "snRNAseq.{CELL_TYPE}.Q{QUANTILE}.{CHR}.annot.gz",
             bfile_folder = REF_DIR + "1000G_EUR_Phase3_plink",
             snps_folder = REF_DIR + "hapmap3_snps"
    output:  ANN_DIR + "snRNAseq.{CELL_TYPE}.Q{QUANTILE}.{CHR}.l2.ldscore.gz"
    conda:   "envs/ldsc.yml"
    params:  bfile = REF_DIR + "1000G_EUR_Phase3_plink/1000G.EUR.QC.{CHR}",
             ldscores = ANN_DIR + "snRNAseq.{CELL_TYPE}.Q{QUANTILE}.{CHR}",
             snps = REF_DIR + "w_hm3.snplist_rsIds"
    message: "Running LDSR Phase 3 for {wildcards.CELL_TYPE} Quantile {wildcards.QUANTILE} CHR {wildcards.CHR}" 
    log:     "logs/LDSR/snRNAseq.{CELL_TYPE}.Q{QUANTILE}.Chr{CHR}_ldsc.log"
    shell:
        "python ldsc/ldsc.py --thin-annot --l2 --bfile {params.bfile} --ld-wind-cm 1 "
        "--annot {input.annot} --out {params.ldscores} --print-snps {params.snps} 2> {log}"

rule partitioned_heritability:
    input:   GWAS = GWAS_DIR + "{GWAS}_hg19_ldsc_ready.sumstats.gz",
             LDSR = expand(rules.ldsr.output, CELL_TYPE = config["RNA_CELL_TYPES"], QUANTILE = range(0,11), CHR = range(1,23))
    output:  PART_HERIT_DIR + "snRNAseq_LDSC_{CELL_TYPE}_Q{QUANTILE}_{GWAS}.results"
    conda:   SCRATCH + "envs/ldsc.yml"
    params:  weights = REF_DIR + "weights_hm3_no_hla/weights.",
             baseline = REF_DIR + "baselineLD_v2.2_1000G_Phase3/baselineLD.",
             frqfile = REF_DIR + "1000G_Phase3_frq/1000G.EUR.QC.",
             LD_anns = ANN_DIR + "snRNAseq.{CELL_TYPE}.Q{QUANTILE}.",
             out_file = PART_HERIT_DIR + "snRNAseq_LDSC_{CELL_TYPE}_Q{QUANTILE}_{GWAS}"
    message: "Running Prt Hrt with {wildcards.CELL_TYPE} Q{wildcards.QUANTILE} and {wildcards.GWAS} GWAS"
    log:     "logs/LDSR/snRNAseq.{CELL_TYPE}.Q{QUANTILE}.{GWAS}_partHerit.log"
    shell:
        "python ldsc/ldsc.py --h2 {input.GWAS} --w-ld-chr {params.weights} "
        "--ref-ld-chr {params.baseline},{params.LD_anns} --overlap-annot "
        "--frqfile-chr {params.frqfile} --out {params.out_file} --print-coefficients 2> {log}"


rule create_partHerit_summary:
    input:   expand(PART_HERIT_DIR + "snRNAseq_LDSC_{CELL_TYPE}_Q{QUANTILE}_{GWAS}.results", CELL_TYPE = config["RNA_CELL_TYPES"], QUANTILE = range(0,11), GWAS = config["LDSC_GWAS"])
    output:  PART_HERIT_DIR + "snRNAseq_LDSC_{CELL_TYPE}_{GWAS}_summary.tsv"
    message: "Creating summary file for {wildcards.CELL_TYPE} and {wildcards.GWAS} GWAS"
    params:  dir = PART_HERIT_DIR
    log:     "logs/LDSR/snRNAseq.{CELL_TYPE}.{GWAS}_partHerit.summary.log"
    shell:
        """
        
        head -1 {params.dir}snRNAseq_LDSC_Cer-RG-1_Q1_SCZ.results > {output} 
        grep L2_1 {params.dir}snRNAseq_LDSC_{wildcards.CELL_TYPE}_Q*_{wildcards.GWAS}.results >> {output}
        
        """

        # Last two rules were run in seprate  test script need to test whole thing is working
        
#rule create_ldsc_plots:
#    input:   RESULTS_DIR + "snRNAseq_LDSC_{CELL_TYPE}_{GWAS}_summary.tsv"
#    output:  rds = RESULTS_DIR + "snRNAseq_LDSC_{CELL_TYPE}_{GWAS}.rds",
#             plot = RESULTS_DIR + "snRNAseq_LDSC_{CELL_TYPE}_{GWAS}.png"
#    message: "Creating ldsc plots for {wildcards.CELL_TYPE} and {wildcards.GWAS} GWAS"
#    log:     "logs/LDSR/snRNAseq.{CELL_TYPE}.{GWAS}_partHerit.plots.log"
#    shell:
#             """
             
#             export R_LIBS_USER=/scratch/c.c1477909/R/library
#             module load libgit2/1.1.0
#             /apps/languages/R/4.0.3/el7/AVX512/gnu-8.1/bin/Rscript --vanilla \
#             scripts/R/snRNAseq_LDSC_create_plots.R {wildcards.CELL_TYPE} {input} {output.rds} {output.plot} 2> {log}

#             """

#rule ldsc_mrkdwn_report_rna:
#    # R produces 5 plots but only tracking the final plot here 
#    input:   expand(RESULTS_DIR + "snRNAseq_LDSC_{CELL_TYPE}_{GWAS}.rds", CELL_TYPE = config["RNA_CELL_TYPES"], GWAS = config["SUMSTATS"])
#    output:  RESULTS_DIR + "snRNAseq_LDSC_report.html"
#    params:  markdown_file = MARKDOWN_DIR + "snRNAseq_LDSC_report.Rmd",
#             out_dir = RESULTS_DIR
#    message: "Creating ldsc group plots for all regions and {wildcards.GWAS} GWAS"
#    log:     "logs/LDSR/snRNAseq.AllRegions.All.GWAS_partHerit.group.plots.log"
#    shell:
#             """
#             export R_LIBS_USER=/scratch/c.c1477909/R/library
#             module load libgit2/1.1.0
#             module load pandoc/2.7.3
#             /apps/languages/R/4.0.3/el7/AVX512/gnu-8.1/bin/Rscript --vanilla \
#             scripts/R/snRNAseq_LDSC_report.R {params.markdown_file} {params.out_dir}  2> {log}
             
#             """
        
        
        
# -------------------------------------------------------------------------------------
# -------------------------------------------------------------------------------------
